### ISSUE TITLE
NaN loss during training process

### ISSUE THREAD

**Issue by consistent1997** [ZhengxueCheng/Learned-Image-Compression-with-GMM-and-Attention#8]
Dr. Cheng, bother you again. I have constructed a train function as your compress function. The difference is that my whole model is based on mean-scale architecture in paper "Joint Autoregressive and Hierarchical Priors for Learned Image Compression". I used your range-coder and the function of calculating likelihoods, which you have annotated in code "network.py". The question is that I have met with NaN loss, have you met with this case in your training debugging? If yes, how do you solve it.

**Comment by ZhengxueCheng** [comment]
> Dr. Cheng, bother you again. I have constructed a train function as your compress function. The difference is that my whole model is based on mean-scale architecture in paper "Joint Autoregressive and Hierarchical Priors for Learned Image Compression". I used your range-coder and the function of calculating likelihoods, which you have annotated in code "network.py". The question is that I have met with NaN loss, have you met with this case in your training debugging? If yes, how do you solve it.

Thanks for your question. I am digging into it to check which part has this numerical issue. Because this codes are not my original codes for CVPR2020, but a re-implementation for CLIC submission at 0.15bpp, so I have not met this problem at 0.15bpp. Could you please tell me your training settings? eg. lambda, MSE or MS-SSIM?

FYI, from my previous experiences on tf 1.9/tf1.13, numerical issues typically comes from GDN or Gaussian mixture model. So I will try to replace GDN with ReLU or make the simga as tf.exp(var), instead of tf.abs(var) in the codes to locate this issue.

**Comment by consistent1997** [comment]
My experiments are on tf1.14. My train MSE and MS-SSIM are the same with yours, but I don't know what is your train loss compositions. Train bpp, one of my train loss compositions, is computed by y_likelihood and z_likelihood, which is the same as Balle's bmshj2018.py. Now I have addressed the numerical issue by adding a lower bound for y_likelihood. Finally, thanks for your comment.

**Comment by ZhengxueCheng** [comment]
> My experiments are on tf1.14. My train MSE and MS-SSIM are the same with yours, but I don't know what is your train loss compositions. Train bpp, one of my train loss compositions, is computed by y_likelihood and z_likelihood, which is the same as Balle's bmshj2018.py. Now I have addressed the numerical issue by adding a lower bound for y_likelihood. Finally, thanks for your comment.

Glad to hear that you have solved it!
FYI, in my codes, bpp = (tf.reduce_sum(tf.log(y_likelihoods)) + tf.reduce_sum(tf.log(z_likelihoods)) )/ (-np.log(2) * num_pixels), so I guess it should be the same as yours. I also use the lower bound 1e-9 for y_likelihoods. 
I closed this issue. If you have other questions, please reopen it. Thanks!
